File: rust/0002.rs
/* file: 0002.rs
 * by  : agaric
 * copy: public domain
 * desc: project euler #2 - "even Fibonacci numbers"
 * lang: rust
 */

fn fib(n: usize) -> Vec<usize> {
    let mut v = vec![1usize, 1];
    for i in 2..(n + 1) {
        let sum = v[i - 1] + v[i - 2];
        v.push(sum);
    }
    v
}

fn main() {
    let f = fib(32); // fib(33) > 4 mil
    println!("{}", f.iter()
             .filter(|&&x| x % 2 == 0)
             .fold(0, |sum, x| sum + x));
}

File: go/0002.go
/* file: 0002.go
 * by  : agaric
 * desc: project euler #2 - "even Fibonacci numbers"
 */

package main

import "fmt"

func fib(n int) []int {
	a := make([]int, n)
	a[0], a[1] = 1, 2
	for i := 2; i < n; i++ {
		a[i] = a[i - 1] + a[i - 2]
	}
	return a
}

func main() {
	sum := 0
	for _, x := range fib(32) { // fib(33) > 4 mil
		if x % 2 == 0 {
			sum += x
		}
	}
	fmt.Println(sum)
}
